### PR TITLE
C++ extension preferences framework

### PR DESCRIPTION
**Pull request by Nebiroth**
Basic framework for C++ prefs in which I created one preference to be sent to clangd as a command line arg. This framework should be used for every preference in this extension's scope.

### CODE REVIEW

**Comment by hexa00** [comment]
Please rebase on master instead of merging it in

**Comment by MarkZ3** [comment]
Can you squash some commits? Some temporary changes don't make much sense to keep in history

File: packages/cpp/src/browser/cpp-keybinding.ts [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+} from "@theia/core/lib/common";
+
+@injectable()
+export class CppKeybindingContext implements KeybindingContext {
+    constructor( @inject(EditorManager) protected readonly editorService: EditorManager) { }
+
+    id = 'cpp.keybinding.context';
+
+    isEnabled(arg?: Keybinding) {
+        return this.editorService && !!this.editorService.activeEditor &&
```
Comment (akosyakov) [inline packages/cpp/src/browser/cpp-keybinding.ts]: - whether this.editorService is undefined check seems to be redundant
- extracting activeEditor to the const will make the code more readable 
- it seems that this code is duplicated in the PR

Comment (Nebiroth) [reply packages/cpp/src/browser/cpp-keybinding.ts]: Haven't looked at what causes the bug with the command palette yet.

Comment (akosyakov) [reply packages/cpp/src/browser/cpp-keybinding.ts]: It should be fine with the latest master. @svenefftinge fixed it

File: packages/cpp/src/common/cpp-preferences.ts [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+            "type": "string",
+            "description": "clangd path for compile_commands.json"
+        },
+        "cpp.clangdPath": {
+            "type": "string",
+            "description": "Literal command to start Clangd with. Can contain any relative or absolute path to clangd executable."
+        },
+        "cpp.clangdCommandArgs": {
+            "type": "string",
+            "description": "Literal command to start Clangd with. Can contain any relative or absolute path to clangd executable."
```
Comment (MarkZ3) [inline packages/cpp/src/common/cpp-preferences.ts]: I think this line was copy pasted and you need to mention "args" here?

File: packages/cpp/src/node/cpp-contribution.ts [hunk 3]
```diff
@@ -5,20 +5,53 @@
...
+        return message;
+    }
+
+    protected forward(clientConnection: IConnection, serverConnection: IConnection): void {
+        super.forward(clientConnection, serverConnection);
+    }
+
+    public start(clientConnection: IConnection): void {
+        let command: any = '';
+        let args: string[] = [];
```
Comment (MarkZ3) [inline packages/cpp/src/node/cpp-contribution.ts]: shouldn't "cpp.clangdCommandArgs" be used somewhere around here?

File: packages/cpp/src/common/cpp-preferences.ts [hunk 4]
```diff
@@ -0,0 +1,59 @@
...
+    }
+}
+
+export interface CppConfiguration {
+    'cpp.clangdCompileCommandsPath': string,
+    'cpp.clangdPath': string
+}
+
+export const defaultCppConfiguration: CppConfiguration = {
+    'cpp.clangdCompileCommandsPath': "-compile-commands-dir=/home/ewilenr/theiaEclipse/eclipse/git/llvm/build",
```
Comment (MarkZ3) [inline packages/cpp/src/common/cpp-preferences.ts]: clangdCompileCommandsPath should default to nothing

File: packages/cpp/src/common/cpp-preferences.ts [hunk 5]
```diff
@@ -0,0 +1,59 @@
...
+    'cpp.clangdPath': "/home/ewilenr/theiaEclipse/eclipse/git/llvm/build/bin/compile_commands.json"
```
Comment (MarkZ3) [inline packages/cpp/src/common/cpp-preferences.ts]: cpp.clangdPath should default to either "clangd" or nothing (then in code nothing == clangd)

File: packages/cpp/src/node/cpp-contribution.ts [hunk 6]
```diff
@@ -5,20 +5,53 @@
...
+        if (this.cppPreferences["cpp.clangdPath"] === "") {
```
Comment (MarkZ3) [inline packages/cpp/src/node/cpp-contribution.ts]: if cpp.clangdPath is empty then use /clangd ? Maybe I'm reading typescript wrong :)

**Pull request merged**